### PR TITLE
feat(gossipsub): Add support for custom connection handling (Mix protocol integration)

### PR DESCRIPTION
**Pull request by AkshayaMani**
This PR implements support in GossipSub for protocols like the [Mix protocol](https://rfc.vac.dev/vac/raw/mix), which require custom peer selection and connection handling when publishing messages.

### Summary of changes
- Adds a new `useCustomConn` flag to `GossipSub.publish` (defaults to `false`).
- When `useCustomConn` is `true`, `publish` uses callbacks registered during GossipSub initialization:
  - `peerSelectionCB`: returns the peers to publish to.
  - `customConnCreationCB`: returns a custom connection for sending the message.
- If these callbacks are not provided and `useCustomConn` is `true`, `publish` returns an error.

### Motivation
This is part of ongoing integration work for the Mix protocol — a custom, stateless libp2p protocol designed for anonymous message routing in peer-to-peer networks.
The callback mechanism is designed to be generic and may be reused by other protocols requiring fine-grained control over peer selection and connection setup.

### Notes
- All changes are backward-compatible; default `publish` behavior is unaffected.
- Code has been formatted using `nph`.

### Related
[VAC RAW RFC: Mix Protocol](https://rfc.vac.dev/vac/raw/mix)

> **Note:** This PR also includes a small unrelated change needed to support the custom connection logic. It does not impact GossipSub behavior directly.

### CODE REVIEW

File: libp2p/protocols/pubsub/pubsubpeer.nim [hunk 1]
```diff
@@ -95,6 +95,17 @@ type
...
+  CustomConnectionCallbacks* = object
+    customConnCreationCB*: proc(
+      destAddr: Option[MultiAddress], destPeerId: PeerId, codec: string
+    ): Connection {.gcsafe, raises: [].}
+    peerSelectionCB*: proc(
+      allPeers: HashSet[PubSubPeer],
+      directPeers: HashSet[PubSubPeer],
+      meshPeers: HashSet[PubSubPeer],
+      fanoutPeers: HashSet[PubSubPeer],
+    ): HashSet[PubSubPeer] {.gcsafe, raises: [].}
```
Comment (richard-ramos) [inline libp2p/protocols/pubsub/pubsubpeer.nim]: To make this `CustomConnectionCallbacks` object easier to read, could you extract the procs to a type? That way The object would look like:
```
CustomConnectionCallbacks* = object
    customConnCreationCB*: someCustomProcTypeHere
    peerSelectionCB*: anotherCustomProcTypeHere
```

File: libp2p/protocols/pubsub/pubsubpeer.nim [hunk 2]
```diff
@@ -356,21 +368,42 @@ proc sendMsgSlow(p: PubSubPeer, msg: seq[byte]) {.async: (raises: [CancelledErro
...
-    let conn = p.sendConn
+proc sendMsg(
+    p: PubSubPeer, msg: seq[byte], useCustomConn: bool = false
+): Future[void] {.async: (raises: []).} =
+  var slowPath = false
+  let (conn, connType) =
+    if useCustomConn and p.customConnCallbacks.isSome:
+      let address =
+        if p.address.isSome:
+          some(p.address.get)
```
Comment (richard-ramos) [inline libp2p/protocols/pubsub/pubsubpeer.nim]: since p.address is already `some`,  couldnt you return p.address directly?

File: libp2p/protocols/pubsub/pubsubpeer.nim [hunk 3]
```diff
@@ -356,21 +368,42 @@ proc sendMsgSlow(p: PubSubPeer, msg: seq[byte]) {.async: (raises: [CancelledErro
...
+        else:
+          none(MultiAddress)
+      (
+        p.customConnCallbacks.get().customConnCreationCB(address, p.peerId, p.codec),
+        "custom connection",
```
Comment (richard-ramos) [inline libp2p/protocols/pubsub/pubsubpeer.nim]: Probably makes sense to use a enum for `custom connection`, `send connection` and `slow path`, instead of relying on strings

File: libp2p/protocols/pubsub/gossipsub.nim [hunk 4]
```diff
@@ -719,55 +719,70 @@ method publish*(
...
-        let
-          bandwidth = (g.parameters.bandwidthEstimatebps) div 8 div 1000
-            # Divisions are to convert it to Bytes per ms TODO replace with bandwidth estimate
-          msToTransmit = max(data.len div bandwidth, 1)
-        max(
-          g.parameters.heartbeatInterval.milliseconds div msToTransmit,
-          g.parameters.dLow,
+  if useCustomConn:
+    if g.customConnCallbacks.isSome:
+      peers.incl(
```
Comment (richard-ramos) [inline libp2p/protocols/pubsub/gossipsub.nim]: Might make sense to add a trace log here to know we're using a customConnCallback (similar to how it's done on line 7333)

File: libp2p/protocols/pubsub/gossipsub.nim [hunk 5]
```diff
@@ -719,55 +719,70 @@ method publish*(
...
+          g.subscribedDirectPeers.getOrDefault(topic),
+          g.mesh.getOrDefault(topic),
+          g.fanout.getOrDefault(topic),
         )
-      else:
-        int.high() # unlimited
+      )
+    else:
+      trace "No custom connection callbacks provided, skipping publish"
+      libp2p_gossipsub_failed_publish.inc()
```
Comment (richard-ramos) [inline libp2p/protocols/pubsub/gossipsub.nim]: This condition should not happen, as it would indicate a bad configuration from the developer. Should we use an assertion here and intentionally crash the program?

File: libp2p/protocols/pubsub/gossipsub.nim [hunk 6]
```diff
@@ -719,55 +719,70 @@ method publish*(
 
   var peers: HashSet[PubSubPeer]
```
Comment (vladopajic) [inline libp2p/protocols/pubsub/gossipsub.nim]: in `publish` method it looks like `peers` are created using two ways based on `useCustomConn`.

```
var peers: HashSet[PubSubPeer]

if useCustomConn
  .... populates peers
else 
  ... populates peers
 
# use peers here

```
because this code is hard to follow could we extract these two ways of creating peers in functions so that we get code like this:

```
var peers: HashSet[PubSubPeer]

if useCustomConn
   peers = g.makePeersForPublishUsingCustomCon(topic)
else 
   peers = g.makePeersForPublishDefault(topic)
 
# use peers here

```

it will make `publish` method much easier to read in future.

**Review by vladopajic — APPROVED** [review]
lgtm in general just needs some tidying up

File: libp2p/protocols/pubsub/pubsubpeer.nim (hunk 1 above)
Comment (AkshayaMani) [reply libp2p/protocols/pubsub/pubsubpeer.nim]: Refactored `CustomConnectionCallbacks` to use named proc types: `CustomConnCreationProc` and `CustomPeerSelectionProc`.

File: libp2p/protocols/pubsub/pubsubpeer.nim (hunk 2 above)
Comment (AkshayaMani) [reply libp2p/protocols/pubsub/pubsubpeer.nim]: Simplified this by returning `p.address` directly

File: libp2p/protocols/pubsub/pubsubpeer.nim (hunk 3 above)
Comment (AkshayaMani) [reply libp2p/protocols/pubsub/pubsubpeer.nim]: Replaced connection type strings with a `ConnectionType` enum (`ctCustom`, `ctSend`, `ctSlow`).

File: libp2p/protocols/pubsub/gossipsub.nim (hunk 4 above)
Comment (AkshayaMani) [reply libp2p/protocols/pubsub/gossipsub.nim]: Added a `trace` log in `makePeersForPublishUsingCustomConn` function to indicate when a custom peer selection callback is being used.

File: libp2p/protocols/pubsub/gossipsub.nim (hunk 5 above)
Comment (AkshayaMani) [reply libp2p/protocols/pubsub/gossipsub.nim]: Added an assert to fail when `useCustomConn` is `true` but `customConnCallbacks` is not set.

File: libp2p/protocols/pubsub/gossipsub.nim (hunk 6 above)
Comment (AkshayaMani) [reply libp2p/protocols/pubsub/gossipsub.nim]: Refactored peer selection logic in `publish` into two helper functions: `makePeersForPublishUsingCustomConn` and `makePeersForPublishDefault`.

**Comment by vladopajic** [comment]
hmm, just realized it would be good to have some tests for this code

**Comment by Ben-PH** [comment]
Relevant context; this PR breaks some things in the [mix]() repo. Don't let that stop any progress on this PR. This context might help paint a picture of the bigger picture.

**Comment by AkshayaMani** [comment]
> hmm, just realized it would be good to have some tests for this code

Good point! We can probably add tests for the peer selection logic (custom path) and also verify behavior when `customConnCallbacks` is missing.
Curious if you had any specific edge cases in mind?

> Relevant context; this PR breaks some things in the mix repo. Don't let that stop any progress on this PR. This context might help paint a picture of the bigger picture.

Yep, aware of that — the changes in the mix repo are minor and can be addressed once this PR is merged. Thanks for flagging it!
